File: 04-stack/00084-Largest_Rectangle_in_Histogram/00084-largest_rectangle_in_histogram.go
// 84: Largest Rectangle in Histogram
// https://leetcode.com/problems/largest-rectangle-in-histogram/

package main

import (
	"fmt"
	"math"
)


// SOLUTION
func largestRectangleArea(heights []int) int {
	var stack []int
	l := len(heights)
	result := 0

	for i:=0; i<=l; i++ {
		var h int
		if i==l {h=0} else {h=heights[i]}

		if len(stack)==0 || h>=heights[stack[len(stack)-1]] {
			stack = append(stack, i)
		} else {
			tp := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			var t int
			if len(stack)==0 {t=i} else {t=i-1-stack[len(stack)-1]}
			result = int(math.Max(float64(result), float64(heights[tp] * t)))
			i--
		}
	}

	return result
}


func main() {
	// INPUT
	heights := []int{2,1,5,6,2,3}

	// OUTPUT
	result := largestRectangleArea(heights)
	fmt.Println(result)
}

File: 04-stack/00084-Largest_Rectangle_in_Histogram/00084-largest_rectangle_in_histogram.java
// 84: Largest Rectangle in Histogram
// https://leetcode.com/problems/largest-rectangle-in-histogram/


import java.util.Stack;


class Solution {
    // SOLUTION
    public int largestRectangleArea(int[] heights) {
        Stack<Integer> stack = new Stack<>();
        int l = heights.length;
        int result = 0;

        for (int i=0; i<=l; i++) {
            int h = i==l ? 0 : heights[i];

            if (stack.isEmpty() || h>=heights[stack.peek()])
                stack.push(i);
            else {
                int tp = stack.pop();
                result = Math.max(result, heights[tp]*(stack.isEmpty() ? i : i-1-stack.peek()));
                i--;

            }
        }

        return result;
    }


    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        int[] heights = {2,1,5,6,2,3};

        // OUTPUT
        var result = o.largestRectangleArea(heights);
        System.out.println(result);
    }
}